cmd/dnsmasq-nanny/main.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/dns/pkg/dns/config"
+	"k8s.io/dns/pkg/dnsmasq"
+)
+
+var (
+	opts = struct {
+		dnsmasqArgs    []string
+		dnsmasqExec    string
+		configDir      string
+		restartDnsmasq bool
+		syncInterval   time.Duration
+	}{
+		dnsmasqExec:    "/usr/sbin/dnsmasq",
+		configDir:      "/etc/k8s/dns/dnsmasq-nanny",
+		restartDnsmasq: false,
+		syncInterval:   10 * time.Second,
+	}
+
+	currentConfig *config.Config
+)
+
+func parseFlags() {
+	opts.dnsmasqArgs = dnsmasq.ExtractDnsmasqArgs(&os.Args)
+
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "Usage of %s:\n", os.Args[0])
+		fmt.Fprintf(os.Stderr, `
+Manages the dnsmasq daemon, handles configuration given by the ConfigMap.
+Any arguments given after "--" will be passed directly to dnsmasq itself.
+
+`)
+		flag.PrintDefaults()
+	}
+
+	flag.StringVar(&opts.dnsmasqExec, "dnsmasqExec", opts.dnsmasqExec,
+		"location of the dnsmasq executable")
+	flag.StringVar(&opts.configDir, "configDir", opts.configDir,
+		"location of the configuration")
+	flag.BoolVar(&opts.restartDnsmasq, "restartDnsmasq",
+		opts.restartDnsmasq,
+		"location of the dnsmasq executable")
+	flag.DurationVar(&opts.syncInterval, "syncInterval",
+		opts.syncInterval,
+		"interval to check for configuration updates")
+	flag.Parse()
+}
+
+func main() {
+	parseFlags()
+	glog.V(0).Infof("opts: %v", opts)
+
+	sync := config.NewFileSync(
+		opts.configDir,
+		opts.syncInterval)
+
+	var err error
+	currentConfig, err = sync.Once()
+	if err != nil {
+		glog.Fatalf("Error getting initial config: %v", err)
+	}
+
+	daemon := &dnsmasq.Nanny{Exec: opts.dnsmasqExec}
+	daemon.Configure(opts.dnsmasqArgs, currentConfig)
+	if err := daemon.Start(); err != nil {
+		glog.Fatalf("Could not start dnsmasq with initial configuration: %v", err)
+	}
+
+	configChan := sync.Periodic()
+
+	for {
```

Comment:
May be better to put the majority logics of main() into pkg/dnsmasq and make it some function calls here?